classes/core/APIRouter.inc.php

```diff
@@ -0,0 +1,140 @@
+<?php
+
+/**
+ * @file classes/core/APIRouter.inc.php
+ *
+ * Copyright (c) 2014-2017 Simon Fraser University
+ * Copyright (c) 2000-2016 John Willinsky
+ * Distributed under the GNU GPL v2. For full terms see the file docs/COPYING.
+ *
+ * @class APIRouter
+ * @ingroup core
+ *
+ * @brief Map HTTP requests to a REST API using the Slim microframework.
+ *
+ * Requests for [index.php]/api are intercepted for site-level API requests,
+ * and requests for [index.php]/{contextPath}/api are intercepted for
+ * context-level API requests.
+ */
+
+import('lib.pkp.classes.core.PKPRouter');
+import('classes.core.Request');
+import('classes.handler.Handler');
+
+class APIRouter extends PKPRouter {
+	/** @var APIHandler */
+	var $_handler;
+
+	/**
+	 * Constructor
+	 */
+	public function __construct() {
+		parent::__construct();
+	}
+
+	/**
+	 * Determines whether this router can route the given request.
+	 * @param $request PKPRequest
+	 * @return boolean true, if the router supports this request, otherwise false
+	 */
+	function supports($request) {
+		if (!isset($_SERVER['PATH_INFO'])) return false;
+		$pathInfoParts = explode('/', trim($_SERVER['PATH_INFO'], '/'));
+
+		if (count($pathInfoParts)>=2 && $pathInfoParts[1] == 'api') {
+			// Context-specific API requests: [index.php]/{contextPath}/api
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Get the API version
+	 * @return string
+	 */
+	function getVersion() {
+		$pathInfoParts = explode('/', trim($_SERVER['PATH_INFO'], '/'));
+		return Core::cleanFileVar(isset($pathInfoParts[2]) ? $pathInfoParts[2] : '');
+	}
+
+	/**
+	 * Get the entity being requested
+	 * @return string
+	 */
+	function getEntity() {
+		$pathInfoParts = explode('/', trim($_SERVER['PATH_INFO'], '/'));
```

Comment:
@NateWr : `journal` is the current implementation. I believe `journal`, `path`, `page` and `op` are the reserved parameters related to `disable_path_info`. I am only suggesting `endpoint` but we could use anything. 

> I wonder if we'll have any name collisions with endpoint arguments. Maybe we can prefix the path info params?

Initially I wanted to go down that road. But I believe this will lead to complex implementation with multiple issues including parameters name collision among other. That's the reason why I suggest we introduce a single parameter (which I called `endpoint`). That parameter's value will be the full endpoint. For instance (`/api/v1/submissions/5` or `/api/v1/submissions/9/files` or `/galleys/12`).  

In the `APIRouter`, I am introducing a new helper function `getPathInfoParts()` that will try to get path info paths depending on `disable_path_info`'s value. That part is now implemented and working. 

The remaining issue is the integration with Slim because Slim framework processes the request behind the scene. Slim can work without url rewrite but expected a specific syntax. Something like `/index.php/<context>/api/v1/submissions/5` (which is similar to ojs url when pathinfo is not disabled). I still need to figure out a workaround. That's where I am currently stuck.

@asmecher